File: javascript/triangle/triangle.js
class Triangle {
    constructor(s1, s2, s3) {
        this.sides = [s1, s2, s3].sort((a, b) => a - b);
    }
    
    kind() {
        // Triangles must have three sides
        if (this.sides.length !== 3) throw new Error("Not a trinagle");
        // Triangle sides must be positive non-zero values
        if (this.sides.some(s => s <= 0)) throw new Error("Not a trinagle");
        // Sum of smaller two sides must be greater than third side
        if (this.sides[0] + this.sides[1] < this.sides[2]) throw new Error("Violates triangle equality");

        if (this.sides[0] === this.sides[1] && this.sides[1] === this.sides[2]) {
            return "equilateral";
        } else if (this.sides[0] === this.sides[1] || this.sides[1] === this.sides[2]) {
            return "isosceles";
        } else {
            return "scalene";
        }
    }
}
  
export default Triangle;

File: go/triangle/triangle.go
// Package triangle provides functions to determine triangle type and validity.
package triangle

import (
	"math"
	"sort"
)

type Kind string

const (
	NaT = "NaT" // not a triangle
	Equ = "Equ" // equilateral
	Iso = "Iso" // isosceles
	Sca = "Sca" // scalene
)

// IsValidLength determines if a value is a proper length value.
func IsValidLength(v float64) bool {
    if !(v > 0) || math.IsInf(v, 1) {
        return false
    }
    return true
}

// KindFromSides determines triangle kind.
func KindFromSides(a, b, c float64) Kind {
	if !IsValidLength(a) || !IsValidLength(b) || !IsValidLength(c) {
        return NaT
	}

	if a == b && b == c {
		return Equ
	}

	sides := []float64{a, b, c}
	sort.Float64s(sides)
	if sides[0]+sides[1] < sides[2] {
		return NaT
	}

	if sides[0] == sides[1] || sides[1] == sides[2] {
		return Iso
	}

	return Sca
}
